// @rocket.chat/apps-engine/server/runtime/AppsEngineVM2Runtime.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppsEngineVM2RuntineTestFixture = void 0;
const alsatian_1 = require("alsatian");
const AppsEngineVM2Runtime_1 = require("./AppsEngineVM2Runtime");
let AppsEngineVM2RuntineTestFixture = class AppsEngineVM2RuntineTestFixture {
    constructor() {
        this.app = {
            getName: () => 'app-name',
        };
    }
    setupFixture() { }
    setup() { }
    runCodeTest(...args) {
        const code = args[0];
        const sandbox = args[1];
        const options = args[2];
        const response = args[3];
        const result = AppsEngineVM2Runtime_1.AppsEngineVM2Runtime.runCode(code, sandbox, options);
        if (result) {
            if (result instanceof Function) {
                alsatian_1.Expect(result()).toBe(response);
            }
        }
    }
    runInSandbox(...args) {
        return __awaiter(this, void 0, void 0, function* () {
            const code = args[0];
            const sandbox = args[1];
            const options = args[2];
            const response = args[3];
            const instance = new AppsEngineVM2Runtime_1.AppsEngineVM2Runtime(this.app, (mod) => mod);
            const result = yield instance.runInSandbox(code, sandbox, options);
            alsatian_1.Expect(result()).toBe(response);
        });
    }
};
__decorate([
    alsatian_1.SetupFixture,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], AppsEngineVM2RuntineTestFixture.prototype, "setupFixture", null);
__decorate([
    alsatian_1.Setup,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], AppsEngineVM2RuntineTestFixture.prototype, "setup", null);
__decorate([
    alsatian_1.TestCase(`module.exports = () => { return 'Hello World First case'};`, { someSandbox: true }, {
        timeout: 10,
        filename: 'filename.ts',
        returnAllExports: true,
    }, 'Hello World First case'),
    alsatian_1.TestCase(`module.exports = { method: () => { return 'Hello World Second case' } }`, null, {
        timeout: 10,
        returnAllExports: false,
    }, 'Hello World Second case'),
    alsatian_1.TestCase(`module.exports = () => { return 'Hello World Third case'};`, { someSandbox: true }, null, 'Hello World Third case'),
    alsatian_1.TestCase(`module.exports = () => { return 'Hello World Fourth case'};`, null, null),
    alsatian_1.TestCase(`module.exports = () => { return 'Hello World Fifth case'};`, { require: () => 'module' }, null, 'Hello World Fourth case'),
    alsatian_1.Test('AppsEngineVM2Runtime.runCode'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], AppsEngineVM2RuntineTestFixture.prototype, "runCodeTest", null);
__decorate([
    alsatian_1.TestCase(`module.exports = () => { return 'Hello World'};`, { someSandbox: true }, {
        timeout: 10,
        filename: 'filename.ts',
        returnAllExports: true,
    }, 'Hello World'),
    alsatian_1.AsyncTest('new AppsEngineVM2Runtime().runInSandbox(code, sandbox, options)'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], AppsEngineVM2RuntineTestFixture.prototype, "runInSandbox", null);
AppsEngineVM2RuntineTestFixture = __decorate([
    alsatian_1.TestFixture('AppsEngineVM2Runtine')
], AppsEngineVM2RuntineTestFixture);
exports.AppsEngineVM2RuntineTestFixture = AppsEngineVM2RuntineTestFixture;
